### PR TITLE
Fog of war

### PR DESCRIPTION
**Pull request by Uppun**

### CODE REVIEW

File: src/Grid.js [hunk 1]
```diff
@@ -31,12 +41,27 @@ export default class Grid {
...
+        if (this.grid[y]) {
+            if (this.grid[y][x]) {
                 return this.grid[y][x];
             }
         }
 
         return null;
     }
+
+    getVisible = (x, y) => {
```
Comment (meishuu) [inline src/Grid.js]: does not need to be bound

File: src/Cell.js [hunk 2]
```diff
@@ -19,6 +19,12 @@ export default class Cell {
         this.grid = grid;
         this.passable = true;
         this.terrain = Cell.Terrain.NORMAL;
+        this.fogVisibility = Cell.fogVisibilityLevels.VISIBLE;
```
Comment (meishuu) [inline src/Cell.js]: thinking about it, i wonder if we can just do something similar to `visibleCells` like a `seenCells` instead of saving this on the cell itself

File: src/Grid.js [hunk 3]
```diff
@@ -11,14 +11,24 @@ export default class Grid {
...
-            for (let j = 0; j < this.width  ; j++) {
+            for (let j = 0; j < this.width; j++) {
                 newGrid.grid[i][j].passable = this.grid[i][j].passable;
             }
         }
 
         return newGrid;
     }
 
+    obfuscateGrid(x, y, grid, visibleCells) {
```
Comment (meishuu) [inline src/Grid.js]: this no longer does any obfuscating anymore :blobcatshy:

File: src/components/CellComponent.js [hunk 4]
```diff
@@ -12,11 +12,12 @@ export default class CellComponent extends Component {
     }
 
     render() {
-        const {passable, type, terrain} = this.props;
+        const {passable, type, terrain, fogVisibility} = this.props;
         const cellClasses = classNames(
             'cell', passable ? 'passable' : 'impassable', 
             {player: type === 'player', end: type === 'end'}, 
-            {mountain: terrain === Cell.Terrain.MOUNTAIN && type === Cell.Terrain.NORMAL && passable}
+            {mountain: terrain === Cell.Terrain.MOUNTAIN && type === Cell.Terrain.NORMAL && passable},
```
Comment (meishuu) [inline src/components/CellComponent.js]: you should just put these all in the same object

File: src/stores/A_store.js [hunk 5]
```diff
@@ -71,6 +71,53 @@ class A_store extends ReduceStore {
...
+                const pathGrid = new Grid(SIZE, SIZE);
+
+                for (let row = 0; row < SIZE; row++) {
+                    for (let column = 0; column < SIZE; column++) {
+                        pathGrid.grid[row][column].fogVisibility = Cell.fogVisibilityLevels.UNKNOWN;
+                    }
+                }
+                let visibleCells = revealedGrid.getVisible(x, y);
+                revealedGrid.obfuscateGrid(x, y, pathGrid, visibleCells);
+                visibleCells = pathGrid.getVisible(x, y);
```
Comment (meishuu) [inline src/stores/A_store.js]: i see why this is jank now, hmm
let's address the other things first and then revisit this thing

File: src/Cell.js [hunk 6]
```diff
@@ -21,6 +21,7 @@ export default class Cell {
         this.terrain = Cell.Terrain.NORMAL;
     }
 
+
```
Comment (meishuu) [inline src/Cell.js]: 🤔

File: src/stores/A_store.js [hunk 7]
```diff
@@ -60,7 +60,7 @@ class A_store extends ReduceStore {
             }
 
             case ActionTypes.RESET: {
-                return {pathGrid: new Grid(SIZE, SIZE), player: {x: -1, y: -1}, end: {x: -2, y: -2}, stage: 'STARTP', path: ''};
+                return {pathGrid: new Grid(SIZE, SIZE), player: {x: -1, y: -1}, end: {x: -2, y: -2}, stage: 'STARTP', path: '', visibleCells: '', seenCells: ''};
```
Comment (meishuu) [inline src/stores/A_store.js]: feeling like this should have newlines now :blobcatdizzy:

also, path and visibleCells and seenCells initialized to a string?

File: src/stores/A_store.js [hunk 8]
```diff
@@ -71,6 +71,53 @@ class A_store extends ReduceStore {
...
+                const start = pathGrid.getCell(x, y);
+                const goal = pathGrid.getCell(end.x, end.y);
+                const path = Astar(start, goal, Cell.heuristic);
+                path.shift();
+
+                return {...state, pathGrid, revealedGrid, path, stage: 'STEP_FOG', visibleCells, seenCells};
+            }
+
+            case ActionTypes.STEP_FOG: {
+                let path = [...state.path];
```
Comment (meishuu) [inline src/stores/A_store.js]: one thing you could do here is
```js
let [nextLocation, ...path] = state.path;
```

File: src/stores/A_store.js [hunk 9]
```diff
@@ -71,6 +71,53 @@ class A_store extends ReduceStore {
...
+                const {revealedGrid, end, seenCells} = state;
+                const nextLocation = path.shift();
+                let stage, pathGrid, visibleCells;
```
Comment (meishuu) [inline src/stores/A_store.js]: nit: usually we discourage declaring multiple variables on the same line (outside of destructuring)

File: src/stores/A_store.js [hunk 10]
```diff
@@ -71,6 +71,53 @@ class A_store extends ReduceStore {
...
 
+            case ActionTypes.GENERATE_FOG: {
+                const {player, end} = state;
+                const {x, y} = player;
+                const revealedGrid = mapGenerator(player, end, new Grid(SIZE, SIZE));
+                const pathGrid = new Grid(SIZE, SIZE);
+
+                let visibleCells = revealedGrid.getVisible(x, y);
+                revealedGrid.obfuscateGrid(x, y, pathGrid, visibleCells);
+                let seenCells;
```
Comment (meishuu) [inline src/stores/A_store.js]: why declare and then assign a couple lines later 🤔

Comment (meishuu) [reply src/stores/A_store.js]: also doesn't look like it's reassigned

File: src/stores/A_store.js [hunk 11]
```diff
@@ -71,6 +71,53 @@ class A_store extends ReduceStore {
                 return {...state, stage, path, player: {x: nextLocation.x, y: nextLocation.y}};
             }
 
+            case ActionTypes.GENERATE_FOG: {
+                const {player, end} = state;
+                const {x, y} = player;
+                const revealedGrid = mapGenerator(player, end, new Grid(SIZE, SIZE));
+                const pathGrid = new Grid(SIZE, SIZE);
+
+                let visibleCells = revealedGrid.getVisible(x, y);
```
Comment (meishuu) [inline src/stores/A_store.js]: okay, so, how about this
instead of doing a `getVisible()` twice, first on `revealedGrid` and second on `pathGrid`, we do it just once – on `pathGrid` since that's the final value we want to use anyway
then we change `obfuscateGrid()` to something like `copyCells()` which takes in a set of cells
so then `revealedGrid.copyCells(visibleCells)` just iterates over the set and copies over terrain/traversable stuff and we save on having to call `getVisible()` twice

File: src/stores/A_store.js [hunk 12]
```diff
@@ -71,6 +71,53 @@ class A_store extends ReduceStore {
...
+                    const start = pathGrid.getCell(nextLocation.x, nextLocation.y);
+                    const goal = pathGrid.getCell(end.x, end.y);
+                    path = Astar(start, goal, Cell.heuristic);
+                    path.shift();
+                } else {
+                    stage = 'RESET';
+                    pathGrid = revealedGrid;
+                }
+
+                return {...state, stage, path, player: {x: nextLocation.x, y: nextLocation.y}, pathGrid, visibleCells, seenCells}
```
Comment (meishuu) [inline src/stores/A_store.js]: i continue to think any object with more than a couple of properties probably ought to be on multiple lines

if you feel strongly about this then go ahead but you do this a ton and imo it makes it very hard to check at a glance what properties are involved

File: src/Grid.js [hunk 13]
```diff
@@ -11,14 +11,23 @@ export default class Grid {
...
+        for (let cell of visibleCells) {
+            grid.grid[cell.y][cell.x].terrain = cell.terrain;
+            grid.grid[cell.y][cell.x].passable = cell.passable;
```
Comment (meishuu) [inline src/Grid.js]: i feel like non-passable cells is really just another terrain type 🤔

File: src/Grid.js (hunk 1 above)
Comment (Uppun) [reply src/Grid.js]: 🐇

File: src/Cell.js (hunk 2 above)
Comment (Uppun) [reply src/Cell.js]: 🐇

File: src/Grid.js (hunk 3 above)
Comment (Uppun) [reply src/Grid.js]: 🐇

File: src/components/CellComponent.js (hunk 4 above)
Comment (Uppun) [reply src/components/CellComponent.js]: 🐇

File: src/Cell.js (hunk 6 above)
Comment (Uppun) [reply src/Cell.js]: 🐇

File: src/stores/A_store.js (hunk 10 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/stores/A_store.js (hunk 9 above)
Comment (Uppun) [reply src/stores/A_store.js]: :rabbit2:

File: src/stores/A_store.js (hunk 7 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/stores/A_store.js (hunk 8 above)
Comment (Uppun) [reply src/stores/A_store.js]: :rabbit2:

File: src/stores/A_store.js (hunk 11 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/stores/A_store.js (hunk 12 above)
Comment (Uppun) [reply src/stores/A_store.js]: 🐇

File: src/Grid.js (hunk 13 above)
Comment (Uppun) [reply src/Grid.js]: 🐇

File: src/Cell.js [hunk 14]
```diff
@@ -56,7 +56,7 @@ export default class Cell {
         for (const offset of neighborOffsets) {
             const neighbor = this.grid.getCell(this.x + offset.coords[0], this.y + offset.coords[1]);
 
-            if (!neighbor || !neighbor.passable) {
+            if (!neighbor || !neighbor.terrain === 'wall') {
```
Comment (meishuu) [inline src/Cell.js]: use constant

File: src/Grid.js [hunk 15]
```diff
@@ -11,14 +11,20 @@ export default class Grid {
...
+            for (let j = 0; j < this.width; j++) {
+                newGrid.grid[i][j].terrain = this.grid[i][j].terrain;
             }
         }
 
         return newGrid;
     }
 
+    copyCells(visibleCells) {
+        for (let cell of visibleCells) {
```
Comment (meishuu) [inline src/Grid.js]: `const`

File: src/components/CellComponent.js [hunk 16]
```diff
@@ -12,11 +12,12 @@ export default class CellComponent extends Component {
...
 
     render() {
-        const {passable, type, terrain} = this.props;
+        const {type, terrain, fogVisibility} = this.props;
         const cellClasses = classNames(
-            'cell', passable ? 'passable' : 'impassable', 
-            {player: type === 'player', end: type === 'end'}, 
-            {mountain: terrain === Cell.Terrain.MOUNTAIN && type === Cell.Terrain.NORMAL && passable}
+            'cell', terrain !== 'wall' ? 'passable' : 'impassable', 
+            {player: type === 'player', end: type === 'end',
```
Comment (meishuu) [inline src/components/CellComponent.js]: style this obj better (one line per property)

File: src/mapGenerator.js [hunk 17]
```diff
@@ -64,13 +64,13 @@ export default function mapGenerator(start, goal, grid) {
             }
         }
 
-        const isHoleSafe = Array.from(borderCells).every(cell => cell.passable);
+        const isHoleSafe = Array.from(borderCells).every(cell => cell.terrain !== 'wall');
```
Comment (meishuu) [inline src/mapGenerator.js]: use constant

File: src/stores/A_store.js [hunk 18]
```diff
@@ -36,8 +36,11 @@ class A_store extends ReduceStore {
                 const {x, y} = action;
                 const pathGrid = state.pathGrid.clone();
                 const cell = pathGrid.getCell(x, y);
-
-                cell.passable = !cell.passable;
+                if (cell.terrain === 'wall') {
```
Comment (meishuu) [inline src/stores/A_store.js]: use constant

File: src/stores/A_store.js [hunk 19]
```diff
@@ -60,15 +63,81 @@ class A_store extends ReduceStore {
             }
 
             case ActionTypes.RESET: {
-                return {pathGrid: new Grid(SIZE, SIZE), player: {x: -1, y: -1}, end: {x: -2, y: -2}, stage: 'STARTP', path: ''};
+                return {pathGrid: new Grid(SIZE, SIZE),
```
Comment (meishuu) [inline src/stores/A_store.js]: generally multiline objs are styled as
```js
{
  k: 'v',
  trailing: 'comma',
}
```

we're already doing this in getInitialState so please be consistent

**Review by meishuu — APPROVED** [review]
looks good aside from style issues

File: src/Cell.js (hunk 14 above)
Comment (Uppun) [reply src/Cell.js]: :rabbit2:

File: src/Grid.js (hunk 15 above)
Comment (Uppun) [reply src/Grid.js]: 🐇

File: src/components/CellComponent.js (hunk 16 above)
Comment (Uppun) [reply src/components/CellComponent.js]: 🐇

File: src/stores/A_store.js (hunk 19 above)
Comment (Uppun) [reply src/stores/A_store.js]: :rabbit2:

File: src/stores/A_store.js (hunk 18 above)
Comment (Uppun) [reply src/stores/A_store.js]: :rabbit2:

File: src/mapGenerator.js (hunk 17 above)
Comment (Uppun) [reply src/mapGenerator.js]: :rabbit2:

File: src/Grid.js [hunk 20]
```diff
@@ -11,14 +11,20 @@ export default class Grid {
...
-                newGrid.grid[i][j].passable = this.grid[i][j].passable;
+            for (let j = 0; j < this.width; j++) {
+                newGrid.grid[i][j].terrain = this.grid[i][j].terrain;
             }
         }
 
         return newGrid;
     }
 
+    copyCells(visibleCells) {
```
Comment (meishuu) [inline src/Grid.js]: nit: `visibleCells` doesn't make sense here
e.g. if we were adding better map-making and added a clone tool, we could copy some arbitrary set of cells; nothing about them made them specifically "visible cells" as opposed to... non-visible cells

File: src/components/CellComponent.js [hunk 21]
```diff
@@ -12,11 +12,15 @@ export default class CellComponent extends Component {
...
     render() {
-        const {passable, type, terrain} = this.props;
+        const {type, terrain, fogVisibility} = this.props;
         const cellClasses = classNames(
-            'cell', passable ? 'passable' : 'impassable', 
-            {player: type === 'player', end: type === 'end'}, 
-            {mountain: terrain === Cell.Terrain.MOUNTAIN && type === Cell.Terrain.NORMAL && passable}
+            'cell', 
+            terrain !== 'wall' ? 'passable' : 'impassable', 
+            {player: type === 'player',
```
Comment (meishuu) [inline src/components/CellComponent.js]: plis follow object styling

File: src/components/CellComponent.js [hunk 22]
```diff
@@ -12,11 +12,15 @@ export default class CellComponent extends Component {
...
         const cellClasses = classNames(
-            'cell', passable ? 'passable' : 'impassable', 
-            {player: type === 'player', end: type === 'end'}, 
-            {mountain: terrain === Cell.Terrain.MOUNTAIN && type === Cell.Terrain.NORMAL && passable}
+            'cell', 
+            terrain !== 'wall' ? 'passable' : 'impassable', 
+            {player: type === 'player', 
+            end: type === 'end', 
+            mountain: terrain === Cell.Terrain.MOUNTAIN && type === 'normal',
+            seen: fogVisibility === 'seen',
```
Comment (meishuu) [inline src/components/CellComponent.js]: one thing we could do is straight pass in `fogVisibility` since `classNames()` correctly handles falsy values and the resulting classname is the same as the value

in that case we may consider turning `fogVisibility` into a generic `className`, or optionally a `fogClassName` if you so desire

File: src/components/GridComponent.js [hunk 23]
```diff
@@ -93,12 +103,23 @@ class PathfinderGrid extends Component {
                                 } else if (end.y === rowIndex && end.x === columnIndex) {
                                   type = 'end';
                                 }
-                                return (<Cell passable={cell.passable} key={columnIndex} type={type} handleClick={this.handleCellClick} x={cell.x} y={cell.y} terrain={cell.terrain} />);
+                                let fogVisibility;
+                                if (visibleCells && seenCells) {
```
Comment (meishuu) [inline src/components/GridComponent.js]: there's a couple of refactoring ideas here **depending on your preference**

1. iirc there's no case where we have only one but not both of `visibleCells` and `seenCells`. we could condense these:
```js
if (visibleCells && !visibleCells.has(cell)) {
  if (seenCells && seenCells.has(cell)) {
    // ...
  } else {
    // ...
  }
}
```
pros: one less layer of indents
cons: longer ifs
(i'd prefer being able to omit the null check but that's up to how you're handling your store)

2. could one-line:
```js
fogVisibility = seenCells.has(cell) ? 'seen' : 'unknown';
```
pros: clearer that it's always set if `!visibleCells.has(cell)`, also you like one-liners
cons: it's one fatter line tho

File: src/components/GridComponent.js [hunk 24]
```diff
@@ -93,12 +103,23 @@ class PathfinderGrid extends Component {
...
+                                    if (!visibleCells.has(cell)) {
+                                        if (seenCells.has(cell)) {
+                                            fogVisibility = 'seen';
+                                        } else {
+                                            fogVisibility = 'unknown';
+                                        }
+                                    }
+                                }
+                                return (<Cell key={columnIndex} type={type} handleClick={this.handleCellClick} x={cell.x} y={cell.y} terrain={cell.terrain} fogVisibility={fogVisibility}/>);
```
Comment (meishuu) [inline src/components/GridComponent.js]: space before `/>` p l e a s e

File: src/stores/A_store.js [hunk 25]
```diff
@@ -36,8 +36,11 @@ class A_store extends ReduceStore {
                 const {x, y} = action;
                 const pathGrid = state.pathGrid.clone();
                 const cell = pathGrid.getCell(x, y);
-
-                cell.passable = !cell.passable;
+                if (cell.terrain === Cell.Terrain.WALL) {
```
Comment (meishuu) [inline src/stores/A_store.js]: out of scope: maybe we should make this cycle through the types now that we have mountains 🤔

File: src/stores/A_store.js [hunk 26]
```diff
@@ -60,15 +63,89 @@ class A_store extends ReduceStore {
             }
 
             case ActionTypes.RESET: {
-                return {pathGrid: new Grid(SIZE, SIZE), player: {x: -1, y: -1}, end: {x: -2, y: -2}, stage: 'STARTP', path: ''};
+                return {
+                    pathGrid: new Grid(SIZE, SIZE), 
+                        player: {x: -1, y: -1},
```
Comment (meishuu) [inline src/stores/A_store.js]: plis why two levels of indent

File: src/stores/A_store.js [hunk 27]
```diff
@@ -60,15 +63,89 @@ class A_store extends ReduceStore {
...
             case ActionTypes.STEP: {
-                const path = [...state.path];
-                const nextLocation = path.shift();
+                const [nextLocation, ...path] = state.path;
 
                 const stage = path.length > 0 ? 'STEP' : 'RESET';
-                return {...state, stage, path, player: {x: nextLocation.x, y: nextLocation.y}};
+                return {
+                    ...state, 
+                        stage,
```
Comment (meishuu) [inline src/stores/A_store.js]: :amegablobsweats:

File: src/stores/A_store.js [hunk 28]
```diff
@@ -60,15 +63,89 @@ class A_store extends ReduceStore {
...
+                revealedGrid.copyCells(visibleCells);
+                const seenCells = pathGrid.getVisible(x, y);
+
+                const start = pathGrid.getCell(x, y);
+                const goal = pathGrid.getCell(end.x, end.y);
+                const path = Astar(start, goal, Cell.heuristic);
+                path.shift();
+
+                return {
+                        ...state,
```
Comment (meishuu) [inline src/stores/A_store.js]: double double 🍔

File: src/stores/A_store.js [hunk 29]
```diff
@@ -60,15 +63,89 @@ class A_store extends ReduceStore {
...
+                    const goal = pathGrid.getCell(end.x, end.y);
+                    path = Astar(start, goal, Cell.heuristic);
+                    path.shift();
+                } else {
+                    stage = 'RESET';
+                    pathGrid = revealedGrid;
+                }
+
+                return {
+                        ...state,
```
Comment (meishuu) [inline src/stores/A_store.js]: at least be consistent within your own changeset :blobcatverysad:

*one level of indent* for the properties, ending brace is on same level as the line with starting brace

again, just look at `getInitialState`

File: src/stores/A_store.js [hunk 30]
```diff
@@ -60,15 +63,89 @@ class A_store extends ReduceStore {
...
+                    revealedGrid.copyCells(visibleCells);
+
+                    for (const cell of visibleCells) {
+                        seenCells.add(cell);
+                    }
+
+                    const start = pathGrid.getCell(nextLocation.x, nextLocation.y);
+                    const goal = pathGrid.getCell(end.x, end.y);
+                    path = Astar(start, goal, Cell.heuristic);
+                    path.shift();
```
Comment (meishuu) [inline src/stores/A_store.js]: so, i'm noticing something here

this is `GENERATE_FOG`:
```js
                const visibleCells = pathGrid.getVisible(x, y);
                revealedGrid.copyCells(visibleCells);

                const seenCells = pathGrid.getVisible(x, y); // same as visibleCells

                const start = pathGrid.getCell(x, y);
                const goal = pathGrid.getCell(end.x, end.y);
                const path = Astar(start, goal, Cell.heuristic);
                path.shift();
```

this is `STEP_FOG`:
```js
                    visibleCells = pathGrid.getVisible(nextLocation.x, nextLocation.y);
                    revealedGrid.copyCells(visibleCells);

                    for (const cell of visibleCells) {
                        seenCells.add(cell);
                    }

                    const start = pathGrid.getCell(nextLocation.x, nextLocation.y);
                    const goal = pathGrid.getCell(end.x, end.y);
                    path = Astar(start, goal, Cell.heuristic);
                    path.shift();
```
feeling like there is some DRY opportunity here

i don't think this is a blocker tho

File: src/Grid.js (hunk 20 above)
Comment (Uppun) [reply src/Grid.js]: :rabbit2:

File: src/components/CellComponent.js (hunk 21 above)
Comment (Uppun) [reply src/components/CellComponent.js]: 🐇

File: src/components/CellComponent.js (hunk 22 above)
Comment (Uppun) [reply src/components/CellComponent.js]: 🐇

File: src/components/GridComponent.js (hunk 23 above)
Comment (Uppun) [reply src/components/GridComponent.js]: 🐇

File: src/stores/A_store.js [hunk 31]
```diff
@@ -36,8 +57,15 @@ class A_store extends ReduceStore {
                 const {x, y} = action;
                 const pathGrid = state.pathGrid.clone();
                 const cell = pathGrid.getCell(x, y);
-
-                cell.passable = !cell.passable;
+                if (cell.terrain === Cell.Terrain.WALL) {
+                    cell.terrain = Cell.Terrain.MOUNTAIN;
+                } else {
+                    if (cell.terrain === Cell.Terrain.MOUNTAIN) {
```
Comment (meishuu) [inline src/stores/A_store.js]: flatten this into an `} else if (...) {`

**Review by meishuu — APPROVED** [review]
juan nit
